synapse/handlers/sso.py

```diff
@@ -172,24 +176,38 @@ async def get_mxid_from_sso(
                 to an additional page. (e.g. to prompt for more information)
 
         """
-        # first of all, check if we already have a mapping for this user
-        previously_registered_user_id = await self.get_sso_user_by_remote_user_id(
-            auth_provider_id, remote_user_id,
-        )
-        if previously_registered_user_id:
-            return previously_registered_user_id
-
-        # Check for grandfathering of users.
-        if grandfather_existing_users:
-            previously_registered_user_id = await grandfather_existing_users()
+        # grab a lock while we try to find a mapping for this user. This seems...
+        # optimistic, especially for implementations that end up redirecting to
+        # interstitial pages.
+        with (await self._mapping_lock.queue(auth_provider_id)):
```

Comment:
I think this can be:

```suggestion
        with await self._mapping_lock.queue(auth_provider_id):
```